java/vector/src/main/java/org/apache/arrow/vector/compression/AbstractCompressionCodec.java

```diff
@@ -0,0 +1,117 @@
+        "The uncompressed buffer size exceeds the integer limit");
+
+    if (uncompressedBuffer.writerIndex() == 0L) {
+      // shortcut for empty buffer
+      ArrowBuf compressedBuffer = allocator.buffer(CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH);
+      compressedBuffer.setLong(0, 0);
+      compressedBuffer.writerIndex(CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH);
+      uncompressedBuffer.close();
+      return compressedBuffer;
+    }
+
+    ArrowBuf compressedBuffer = doCompress(allocator, uncompressedBuffer);
+    long compressedLength = compressedBuffer.writerIndex() - CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH;
+    long uncompressedLength = uncompressedBuffer.writerIndex();
+
+    if (compressedLength > uncompressedLength) {
+      // compressed buffer is larger, send the raw buffer
+      compressedBuffer.close();
+      compressedBuffer = CompressionUtil.packageRawBuffer(allocator, uncompressedBuffer);
+    } else {
+      writeUncompressedLength(compressedBuffer, uncompressedLength);
+    }
+
+    uncompressedBuffer.close();
+    return compressedBuffer;
+  }
+
+  @Override
+  public ArrowBuf decompress(BufferAllocator allocator, ArrowBuf compressedBuffer) {
+    Preconditions.checkArgument(compressedBuffer.writerIndex() <= Integer.MAX_VALUE,
+        "The compressed buffer size exceeds the integer limit");
+
+    Preconditions.checkArgument(compressedBuffer.writerIndex() >= CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH,
+        "Not enough data to decompress.");
+
+    long decompressedLength = compressedBuffer.getLong(0);
+    if (!MemoryUtil.LITTLE_ENDIAN) {
+      decompressedLength = Long.reverseBytes(decompressedLength);
+    }
+
+    if (decompressedLength == 0L) {
+      // shortcut for empty buffer
+      compressedBuffer.close();
+      return allocator.getEmpty();
+    }
+
+    if (decompressedLength == CompressionUtil.NO_COMPRESSION_LENGTH) {
+      // no compression
+      return CompressionUtil.extractUncompressedBuffer(compressedBuffer);
+    }
+
+    ArrowBuf decompressedBuffer = doDecompress(allocator, compressedBuffer);
+    compressedBuffer.close();
+    return decompressedBuffer;
+  }
+
+  protected void writeUncompressedLength(ArrowBuf compressedBuffer, long uncompressedLength) {
+    if (!MemoryUtil.LITTLE_ENDIAN) {
+      uncompressedLength = Long.reverseBytes(uncompressedLength);
+    }
+    // first 8 bytes reserved for uncompressed length, according to the specification
+    compressedBuffer.setLong(0, uncompressedLength);
+  }
+
+  protected long readUncompressedLength(ArrowBuf compressedBuffer) {
+    long decompressedLength = compressedBuffer.getLong(0);
+    if (!MemoryUtil.LITTLE_ENDIAN) {
+      decompressedLength = Long.reverseBytes(decompressedLength);
+    }
+    return decompressedLength;
+  }
+
+  /**
+   * The method that actually performs the data compression.
+   * Please note that the first 8 bytes of the compressed buffer is reserved for the uncompressed size.
```

Comment:
maybe clarify that is expected that the length of the returned buffer is always "compressed data size in bytes" + 8 bytes reserved at the beginning of the buffer.  

Can you clarify contract for ownership of uncompressed buffer?

Reply:
I have changed the javadoc to clarify the layout, and made the ownership clear. Thanks for the good suggestion.